EasyCommands.Tests/ScriptTests/ScriptTest.cs

```diff
@@ -0,0 +1,182 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using static IngameScript.Program;
+
+namespace EasyCommands.Tests.ScriptTests
+{
+    /// <summary>
+    /// Represents an instance of a ScriptTest. Constructing a new ScriptTest is akin to setting
+    /// up the test, as it instantiates required objects and performs basic setup.
+    /// 
+    /// This implements <see cref="IDisposable"/>, so any post-test cleanup is performed
+    /// upon destruction of the class. Please use this inside of a using statement.
+    /// </summary>
+    class ScriptTest : IDisposable
+    {
+        private GenericListDictionary mockedBlocksByName;
+        private List<IMyBlockGroup> mockedGroups;
+        private Mock<IMyGridTerminalSystem> gridMock;
+        private Program program;
+
+        public List<String> Logger { get; private set; }
+
+        public ScriptTest(String script)
+        {
+            mockedBlocksByName = new GenericListDictionary();
+            mockedGroups = new List<IMyBlockGroup>();
+            Logger = new List<String>();
+
+            // Init static fields that may have been modified from other tests, prepare for testing
+            Program.STATE = ProgramState.STOPPED;
+            Program.LOG_LEVEL = Program.LogLevel.SCRIPT_ONLY;
+            // Setup the CUSTOM_DATA to return the given script
+            // And other required config for mocking
+            gridMock = new Mock<IMyGridTerminalSystem>();
+            var me = new Mock<IMyProgrammableBlock>();
+            MDKFactory.ProgramConfig config = default;
+            config.GridTerminalSystem = gridMock.Object;
+            config.ProgrammableBlock = me.Object;
+            config.Echo = (message) => Logger.Add(message);
+            program = MDKFactory.CreateProgram<Program>(config);
+            me.Setup(b => b.CustomData).Returns(script);
+
+            // Default behavior for broadcast messages
+            // TODO: Replace this with mock objects passed to config setup in Program
+            program.broadcastMessageProvider = () => new List<MyIGCMessage>();
+
+            // Default behavior for locating blocks on the grid
+            gridMock.Setup(g => g.GetBlockGroups(It.IsAny<List<IMyBlockGroup>>(), It.IsAny<Func<IMyBlockGroup, bool>>()))
+                .Callback((Action<List<IMyBlockGroup>, Func<IMyBlockGroup, bool>>)GetBlockGroupsCallback);
+
+            // TODO: Handle custom logic in TextSurfaceHandler
+        }
+
+        /// <summary>
+        /// Run the script until the program reaches the specified state, or until a limited
+        /// number of runs has been reached. The default number of runs before exit is 100.
+        /// </summary>
+        /// <param name="state">The desired end-state.</param>
+        /// <param name="runLimit">The limited number of runs. Defaults to 100</param>
+        public void RunUntil(ProgramState state, int runLimit = 100)
```

Comment:
Thoughts on passing a delegate "checker" to this instead of using runLimit (you could still pass runLimit == 100 as the default)?

It'd be kind of nice to say "run until X condition", sort of like breakpoints in the script to then check assertions of the state of the world.

Reply:
Done! The predicate currently accepts the ScriptTest as its input and I figure we can expose aspects of the program through properties on the ScriptTest.